hasktorch/src/Torch/NN.hs

```diff
@@ -107,9 +112,16 @@ data LinearSpec = LinearSpec { in_features :: Int, out_features :: Int }
 
 data Linear = Linear { weight :: Parameter, bias :: Parameter } deriving (Show, Generic)
 
+linear' :: Linear -> Tensor -> Tensor
+linear' layer input = linear input w b
+    where
+        linear input weight bias = unsafePerformIO $ (cast3 ATen.linear_ttt) input weight bias
+        w = toDependent (weight layer)
+        b = toDependent (bias layer)
+
 instance Randomizable LinearSpec Linear where
   sample LinearSpec{..} = do
-      w <- makeIndependent =<< randn' [in_features, out_features]
+      w <- makeIndependent =<< randn' [out_features, in_features]
```

Comment:
This looks to be a dimension bug in the existing implementation, but please check this when we get around to reviews for the PR since it effects every use of untyped FC layers.

Reply:
Our implemented linear function is `linear input depWeight depBias= matmul input depWeight + depBias`.

We begun to use Pytorch's linear.
Pytorch's linear behaves below.

```
> linear (ones' [10,2]) (ones' [3,2]) (ones' [])
Tensor Float [10,3] [..]
```

Pytorch's linear may be below.
```
linear input depWeight depBias= matmul input (transpose depWeight) + depBias`
```

So this depends on which linear we use.

Reply:
Yes that's what I saw as well. Fair it's more of a difference in convention than bug, but I think it's going to be very confusing if we don't follow the defacto standard. Also, checking into the typed code, it uses the same (pytorch) convention.

Note that there are several untyped examples that re-implement the above linear function, I've changed all occurrences of them as well (not just mnist). They have to be consistent since they're sharing the typeclass implementation.

Reply:
Ouch, we should just decide on a single convention for linear transforms. Otherwise it's going to be too error prone.